### ISSUE TITLE
Error building for deaktop.mac from windows, any idea?

### ISSUE THREAD

**Issue by syrys** [NathanWalker/angular-seed-advanced#246]
Hey Guys,

Im not 100% sure on this thus the post here. Can we build a mac app (`npm run build.desktop.mac`) using windows? I develop in windows, and i just tried to build a mac app of my project for the first time, and it gave me the following error, and wasent sure if i needed to be using a Mac computer to be able to build for mac?

Command used: `npm run build.desktop.mac`
OS: Windows 10
Other: windows/linux builds work just fine (well, builds without error, i havent tested them yet)
Error:

```
....
....
....
[16:06:01] Starting 'desktop.package.mac'...
(node:14728) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it
to a more recent version.
[16:06:25] 'desktop.package.mac' errored after 24 s
[16:06:25] Error: EPERM: operation not permitted, symlink 'Versions\Current\Electron Framework' -> '__PROJECT_DIR__\de
sktop\mac\angular2-seed-advanced.app\Contents\Frameworks\Electron Framework.framework\Electron Framework'
    at Error (native)
[16:06:25] 'desktop.mac' errored after 41 s
[16:06:25] Error in plugin 'run-sequence(desktop.package.mac)'
Message:
    EPERM: operation not permitted, symlink 'Versions\Current\Electron Framework' -> '__PROJECT_DIR__\desktop\mac\angu
lar2-seed-advanced.app\Contents\Frameworks\Electron Framework.framework\Electron Framework'
Details:
    errno: -4048
    code: EPERM
    syscall: symlink
    path: Versions\Current\Electron Framework
    dest: __PROJECT_DIR__\desktop\mac\angular2-seed-advanced.app\Contents\Frameworks\Electron Framework.framework\Elec
tron Framework
Stack:
Error: EPERM: operation not permitted, symlink 'Versions\Current\Electron Framework' -> '__PROJECT_DIR__\desktop\mac\a
ngular2-seed-advanced.app\Contents\Frameworks\Electron Framework.framework\Electron Framework'
    at Error (native)

npm ERR! Windows_NT 10.0.10586
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "build.
desktop.mac"
npm ERR! node v6.3.1
npm ERR! npm  v3.10.3
npm ERR! code ELIFECYCLE
npm ERR! angular2-seed-advanced@0.0.0 build.desktop.mac: `rimraf desktop/mac/* && gulp desktop.mac`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the angular2-seed-advanced@0.0.0 build.desktop.mac script 'rimraf desktop/mac/* && gulp desktop.mac'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the angular2-seed-advanced package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     rimraf desktop/mac/* && gulp desktop.mac
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs angular2-seed-advanced
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls angular2-seed-advanced
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     __PROJECT_DIR__\npm-debug.log
```

**Comment by kbrandwijk** [comment]
Did you try to run the console as administrator? Symlink permission errors seem to be related to that a lot on Windows. Those are only used for darwin packages (using gulp-symdest).

**Comment by syrys** [comment]
@kbrandwijk thanks, running the command on admin mode command prompt worked just fine.

After building, i realised that the app is 150+MB (windows and mac). Why is this the case? Do we really have 150MB worth of data in this project? or is this compiling some un-necessary libraries into the project?

To top it off, i used the admin mode cmd to build the mac app from within windows. Then i tried to zip up the `/desktop/mac` folder using 7zip, but i kept getting errors while zipping saying access denied to certain folders (`frameworks` folder for example). What is going on here?

**Comment by kbrandwijk** [comment]
You can't 7zip the Mac OS output folder on Windows, because that doesn't preserve the symlinks. Also, objects in the mirror may seem larger than they appear, because the symlinks appear as duplicate files and are counted more than once on Windows. Compare the build size to the other platforms, they should be smaller. The electron framework is by far the biggest part, when I do a Mac build on Linux, the electron framework folder is 109 MB.

**Comment by syrys** [comment]
@kbrandwijk thanks for the response. i see. For a simple app, 100+MB still feels pretty HUGE. May i ask what these symlinks are used for? Is there any way i can use/build this without symlinks? TBH my requirements only need angular2 and electron, all other features/integrations from this repo are redundant for me (i just like the tidy setup and quick updates).

Also, if i want to build and publish a mac app (upload to my website or send to a friend) from within windows, how would i go about that?

**Comment by kbrandwijk** [comment]
@syrys There's a discussion going on here for almost two years, as of yet there does not seem to be a clear solution: https://github.com/electron/electron/issues/673. Every app packs the entire electron binaries, including chromium etc., hence the 100 MB. 
If you want to package your app for Mac on Windows, I believe you can zip it from within gulp following the directions described here: https://github.com/joaomoreno/gulp-atom-electron. Update desktop.package.mac to add the gulp-vinyl-zip step. This roughly cuts the package size in half.

**Comment by syrys** [comment]
@kbrandwijk Awesome, i managed to do what you said and it zipped up the mac app (shows up as a 50MB file after zipped). There were no errors on the command line. I then sent the resulting zip file to someone using OSX (which ever the latest one is), and when they unzipped and double clicked the app, they get an error: "APP_NAME cannot be opened because of a problem." 

Any ideas?

**Comment by syrys** [comment]
rest of the error msg on OSX:
**APP_NAME cannot be opened because of a problem**

```
Check with the developer to make sure APP_NAME works with this version of OS X. You may need to reinstall the application. Be sure to install any available updates for the application and OS X.

Click Report to see more detailed information and send a report to Apple.
```

**Comment by syrys** [comment]
just to add to the above error, if i look at more details, i see the following:

```
Termination Reason:    DYLD, [0x1] Library missing

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Library not loaded: @rpath/Electron Framework.framework/Electron Framework
  Referenced from: /Users/USER/Desktop/*/APP_NAME.app/Contents/MacOS/Electron
  Reason: image not found
```

**Comment by kbrandwijk** [comment]
TBH, I have no idea. I do know it's related to the electron build process (gulp-atom-electron and the underlying electron-builder) and not specifically to angular2-seed-advanced. I suggest you have a look at the issues over there, because there are quite a few open issues related to Mac builds on Windows.

**Comment by GoldenTao** [comment]
I saw a similar issue with a fresh Electron app using Electron-Packager. From what I understand (I'm fairly new to Electron) the symlinks created when building don't correct carry over to Mac when you move the app over (or zip the app folder). 

I found I was able to run the app directly from a Windows shared folder, but once I moved the app to the Mac it wouldn't work. 

In the end, the only way I could get it work was to build it directly on the Mac (in a time frame I was happy to spend getting it work).
